File: backend/app/api/v1/endpoints/billing.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.deps import db_session, get_request_context_no_quota
from app.billing.plan import PLANS
from app.billing.service import billing_service
from app.core.runtime import AppError
from app.schemas.billing import BillingCheckoutRequest, BillingCheckoutResponse, BillingConfirmPaymentRequest, BillingOrderRead


router = APIRouter()


@router.get("/billing/plans")
def get_billing_plans():
    return {
        "plans": [
            {
                "plan_name": plan_name,
                "task_limit": config["task_limit"],
                "api_limit": config["api_limit"],
                "price_cents": config["price_cents"],
                "display_price": config["display_price"],
                "allowed_ai_providers": config.get("allowed_ai_providers", []),
                "allowed_ai_models": config.get("allowed_ai_models", []),
                "ai_policy_note": config.get("ai_policy_note", ""),
                "supports_custom_model": config.get("supports_custom_model", False),
            }
            for plan_name, config in PLANS.items()
        ]
    }


@router.get("/billing/current")
def get_current_billing_status(
    db: Session = Depends(db_session),
    auth_context=Depends(get_request_context_no_quota),
):
    subscription = billing_service.get_or_create_subscription(db, workspace_id=auth_context.workspace_id)
    plan = PLANS.get(subscription.plan_name, PLANS["free"])
    return {
        "workspace_id": auth_context.workspace_id,
        "plan_name": subscription.plan_name,
        "status": subscription.status,
        "updated_at": subscription.updated_at.isoformat(),
        "allowed_ai_providers": plan.get("allowed_ai_providers", []),
        "allowed_ai_models": plan.get("allowed_ai_models", []),
        "ai_policy_note": plan.get("ai_policy_note", ""),
        "supports_custom_model": plan.get("supports_custom_model", False),
    }


@router.post("/billing/checkout", response_model=BillingCheckoutResponse)
def create_checkout_order(
    payload: BillingCheckoutRequest,
    db: Session = Depends(db_session),
    auth_context=Depends(get_request_context_no_quota),
):
    order, payment_ready, payment_payload = billing_service.create_checkout_order(
        db,
        workspace_id=auth_context.workspace_id,
        user_id=auth_context.user_id,
        plan_name=payload.plan_name,
        provider_name=payload.provider_name,
    )
    return BillingCheckoutResponse(
        order=BillingOrderRead.model_validate(order),
        payment_ready=payment_ready,
        payment_message="支付订单已创建。" if payment_ready else "支付订单已创建，等微信支付商户参数配置完成后就能发起真实扣款。",
        payment_payload=payment_payload,
    )


@router.get("/billing/orders")
def get_billing_orders(
    db: Session = Depends(db_session),
    auth_context=Depends(get_request_context_no_quota),
):
    orders = billing_service.list_orders(db, workspace_id=auth_context.workspace_id)
    return {"orders": [BillingOrderRead.model_validate(item).model_dump(mode="json") for item in orders]}


@router.post("/billing/orders/{order_id}/confirm")
def confirm_billing_order(
    order_id: int,
    payload: BillingConfirmPaymentRequest,
    db: Session = Depends(db_session),
    auth_context=Depends(get_request_context_no_quota),
):
    raise AppError(
        "PAYMENT_CONFIRM_DISABLED",
        "手动确认支付接口已关闭。现在只能通过微信支付回调通知更新订单状态。",
        "billing",
        403,
    )


@router.post("/billing/orders/{order_id}/fail")
def fail_billing_order(
    order_id: int,
    payload: BillingConfirmPaymentRequest,
    db: Session = Depends(db_session),
    auth_context=Depends(get_request_context_no_quota),
):
    order = billing_service.mark_order_failed(db, order_id=order_id, note=payload.note)
    if order.workspace_id != auth_context.workspace_id:
        return {"success": False, "message": "订单不属于当前工作区"}
    return {
        "success": True,
        "message": "订单已更新为失败状态",
        "order": BillingOrderRead.model_validate(order).model_dump(mode="json"),
    }

File: frontend/lib/api/billing.ts
import { getApiV1BaseUrl } from "@/lib/api-gateway";

function buildAuthHeaders(token?: string) {
  const headers: Record<string, string> = {};
  if (token) {
    headers.Authorization = `Bearer ${token}`;
  }
  return headers;
}

async function readErrorMessage(response: Response, fallback: string) {
  try {
    const data = await response.json();
    return data.message || data.detail || fallback;
  } catch {
    return fallback;
  }
}

export type BillingPlan = {
  plan_name: string;
  task_limit: number;
  api_limit: number;
  price_cents: number;
  display_price: string;
  allowed_ai_providers: string[];
  allowed_ai_models: string[];
  ai_policy_note: string;
  supports_custom_model: boolean;
};

export type CurrentBillingStatus = {
  workspace_id: number;
  plan_name: string;
  status: string;
  updated_at: string;
  allowed_ai_providers: string[];
  allowed_ai_models: string[];
  ai_policy_note: string;
  supports_custom_model: boolean;
};

export type BillingOrder = {
  id: number;
  workspace_id: number;
  user_id: number | null;
  plan_name: string;
  status: string;
  amount_cents: number;
  currency: string;
  provider_name: string | null;
  external_order_id: string | null;
  note: string | null;
  created_at: string;
  updated_at: string;
};

export type BillingCheckoutResponse = {
  order: BillingOrder;
  payment_ready: boolean;
  payment_message: string;
  payment_payload: Record<string, unknown>;
};

export async function getBillingPlans(): Promise<{ plans: BillingPlan[] }> {
  const apiV1 = getApiV1BaseUrl();
  if (!apiV1) {
    throw new Error("当前没有可用的后端地址，套餐信息无法读取");
  }
  const response = await fetch(`${apiV1}/billing/plans`, {
    cache: "no-store",
  });
  if (!response.ok) {
    throw new Error(await readErrorMessage(response, "读取套餐信息失败"));
  }
  return response.json();
}

export async function getCurrentBillingStatus(token?: string): Promise<CurrentBillingStatus> {
  const apiV1 = getApiV1BaseUrl();
  if (!apiV1) {
    throw new Error("当前没有可用的后端地址，当前套餐无法读取");
  }
  const response = await fetch(`${apiV1}/billing/current`, {
    cache: "no-store",
    headers: {
      ...buildAuthHeaders(token),
    },
  });
  if (!response.ok) {
    throw new Error(await readErrorMessage(response, "读取当前套餐失败"));
  }
  return response.json();
}

export async function createBillingCheckoutOrder(
  planName: string,
  providerName: "wechat_pay" | "alipay",
  token?: string
): Promise<BillingCheckoutResponse> {
  const apiV1 = getApiV1BaseUrl();
  if (!apiV1) {
    throw new Error("当前没有可用的后端地址，暂时无法创建支付订单");
  }
  const response = await fetch(`${apiV1}/billing/checkout`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...buildAuthHeaders(token),
    },
    body: JSON.stringify({
      plan_name: planName,
      provider_name: providerName,
    }),
  });
  if (!response.ok) {
    throw new Error(await readErrorMessage(response, "创建支付订单失败"));
  }
  return response.json();
}

export async function getBillingOrders(token?: string): Promise<{ orders: BillingOrder[] }> {
  const apiV1 = getApiV1BaseUrl();
  if (!apiV1) {
    throw new Error("当前没有可用的后端地址，订单记录无法读取");
  }
  const response = await fetch(`${apiV1}/billing/orders`, {
    cache: "no-store",
    headers: {
      ...buildAuthHeaders(token),
    },
  });
  if (!response.ok) {
    throw new Error(await readErrorMessage(response, "读取订单记录失败"));
  }
  return response.json();
}

export async function confirmBillingOrder(orderId: number, token?: string) {
  const apiV1 = getApiV1BaseUrl();
  if (!apiV1) {
    throw new Error("当前没有可用的后端地址，订单确认失败");
  }
  const response = await fetch(`${apiV1}/billing/orders/${orderId}/confirm`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...buildAuthHeaders(token),
    },
    body: JSON.stringify({}),
  });
  if (!response.ok) {
    throw new Error(await readErrorMessage(response, "确认订单失败"));
  }
  return response.json();
}
